File: src/bin/profit_watch.rs
//! Profitability watcher — proves the WINNING path with zero cost. Loops over
//! live mainnet: refresh both pools, build the guarded arb (both directions)
//! from fresh state, simulateTransaction. Most iterations show the guard
//! reverting at leg 2 (no edge). The instant a real edge appears — standing or
//! transient — the sim comes back clean (err=null), meaning a profitable arb
//! exists right now and our tx would land. Logs every clean hit + the spot edge
//! to profit_watch.jsonl. No money, no submission — pure measurement.
//!
//! Usage: RPC_ENDPOINT=<url> ALT_ADDRESS=<alt> [BORROW_USDC=500] [POLL_MS=800] \
//!   [RUN_DIR=runs] cargo run --release --bin profit_watch

use arb_engine::arb::{build_arb_tx, load_alt, PoolData};
use arb_engine::pools::{orca_price, pair, ray_clmm_price};
use base64::Engine;
use solana_hash::Hash;
use solana_pubkey::Pubkey;
use std::fs::OpenOptions;
use std::io::Write;
use std::str::FromStr;
use std::time::Duration;

fn rpc(endpoint: &str, body: serde_json::Value) -> Option<serde_json::Value> {
    for attempt in 0..3 {
        if let Ok(resp) = ureq::post(endpoint).send_json(body.clone()) {
            if let Ok(v) = resp.into_json::<serde_json::Value>() {
                return Some(v);
            }
        }
        std::thread::sleep(Duration::from_millis(200 << attempt));
    }
    None
}

fn account_data(endpoint: &str, addr: &str) -> Option<Vec<u8>> {
    let v = rpc(endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"getAccountInfo","params":[addr,{"encoding":"base64"}]}))?;
    base64::engine::general_purpose::STANDARD.decode(v["result"]["value"]["data"][0].as_str()?).ok()
}

fn now() -> u64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs()
}

fn main() {
    let _ = dotenvy::dotenv();
    let _ = rustls::crypto::ring::default_provider().install_default();
    let endpoint = std::env::var("RPC_ENDPOINT").expect("RPC_ENDPOINT");
    let alt_addr = std::env::var("ALT_ADDRESS").expect("ALT_ADDRESS");
    let borrow_ui: f64 = std::env::var("BORROW_USDC").ok().and_then(|s| s.parse().ok()).unwrap_or(500.0);
    let poll_ms: u64 = std::env::var("POLL_MS").ok().and_then(|s| s.parse().ok()).unwrap_or(800);
    let run_dir = std::env::var("RUN_DIR").unwrap_or_else(|_| "runs".into());
    let borrow_amount = (borrow_ui * 1e6) as u64;
    let cfg = pair();
    // Placeholder signer — simulate with sigVerify=false, so no keypair needed.
    let signer = Pubkey::from_str("Anu6Awu4kxaEDrg1nkpcikx6tJ2xhfVci5TvDrZBsZEB").unwrap();

    let alt = load_alt(&alt_addr, &account_data(&endpoint, &alt_addr).expect("ALT"));
    let _ = std::fs::create_dir_all(&run_dir);
    let out = format!("{run_dir}/profit_watch.jsonl");

    eprintln!("profit-watch {} borrow {borrow_ui} USDC poll {poll_ms}ms — simulating both dirs; logs clean hits → {out}", cfg.label);
    let (mut iters, mut clean, mut best_edge_bps) = (0u64, 0u64, f64::MIN);

    loop {
        iters += 1;
        let (Some(o), Some(r)) = (account_data(&endpoint, &cfg.orca_pool), account_data(&endpoint, &cfg.ray_pool)) else {
            std::thread::sleep(Duration::from_millis(poll_ms));
            continue;
        };
        // Spot edge for context (stale-free here: just-fetched pools).
        let edge_bps = match (orca_price(&o), ray_clmm_price(&r)) {
            (Some(po), Some(pr)) if po > 0.0 && pr > 0.0 => {
                ((pr - po).abs() / po.min(pr)) * 1e4 - cfg.round_trip_fee_bps()
            }
            _ => f64::NAN,
        };
        if edge_bps.is_finite() && edge_bps > best_edge_bps {
            best_edge_bps = edge_bps;
        }
        let pools = PoolData { orca: o, ray: r };
        let bh = Hash::default();

        for orca_first in [false, true] {
            let Ok(tx) = build_arb_tx(&pools, signer, &alt, borrow_amount, orca_first, None, 0, 10_000, bh) else { continue };
            let b64 = base64::engine::general_purpose::STANDARD.encode(bincode::serialize(&tx).unwrap());
            let v = rpc(&endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"simulateTransaction",
                "params":[b64,{"encoding":"base64","sigVerify":false,"replaceRecentBlockhash":true}]}));
            let val = v.map(|v| v["result"]["value"].clone()).unwrap_or_default();
            if val["err"].is_null() && !val.is_null() {
                clean += 1;
                let dir = if orca_first { "orca→ray" } else { "ray→orca" };
                let cu = val["unitsConsumed"].clone();
                eprintln!("🎉 CLEAN SIM [{dir}] — profitable arb exists NOW, tx would land (edge≈{edge_bps:.2}bp, cu={cu})");
                let row = serde_json::json!({"t": now(), "dir": dir, "edge_bps": edge_bps, "cu": cu, "borrow_usdc": borrow_ui});
                if let Ok(mut f) = OpenOptions::new().create(true).append(true).open(&out) {
                    let _ = writeln!(f, "{row}");
                }
            }
        }
        if iters % 50 == 0 {
            eprintln!("[profit-watch] iters={iters} clean_sims={clean} best_edge={best_edge_bps:.2}bp (need >0 to profit)");
        }
        std::thread::sleep(Duration::from_millis(poll_ms));
    }
}

File: src/bin/profitWatch.ts
// Port of src/bin/profit_watch.rs
//
// Profitability watcher — proves the WINNING path with zero cost. Loops over
// live mainnet: refresh both pools, build the guarded arb (both directions)
// from fresh state, simulateTransaction. Most iterations show the guard
// reverting at leg 2 (no edge). The instant a real edge appears — standing or
// transient — the sim comes back clean (err=null), meaning a profitable arb
// exists right now and our tx would land. Logs every clean hit + the spot edge
// to profit_watch.jsonl. No money, no submission — pure measurement.
//
// Usage: RPC_ENDPOINT=<url> ALT_ADDRESS=<alt> [BORROW_USDC=500] [POLL_MS=800] \
//   [RUN_DIR=runs] tsx src/bin/profitWatch.ts

import 'dotenv/config';
import { mkdirSync } from 'node:fs';
import { appendFile } from 'node:fs/promises';
import { PublicKey } from '@solana/web3.js';
import { buildArbTx, loadAlt, type PoolData } from '../lib/arb.js';
import { orcaPrice, pair, rayClmmPrice } from '../lib/pools.js';

// solana_hash::Hash::default() (32 zero bytes) base58-encodes to this string.
const DEFAULT_BLOCKHASH = '11111111111111111111111111111111';

async function rpc(endpoint: string, body: unknown): Promise<any | undefined> {
  for (let attempt = 0; attempt < 3; attempt++) {
    try {
      const resp = await fetch(endpoint, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(body),
      });
      return await resp.json();
    } catch {
      // fall through to retry
    }
    await new Promise((r) => setTimeout(r, 200 << attempt));
  }
  return undefined;
}

async function accountData(endpoint: string, addr: string): Promise<Buffer | undefined> {
  const v = await rpc(endpoint, {
    jsonrpc: '2.0',
    id: 1,
    method: 'getAccountInfo',
    params: [addr, { encoding: 'base64' }],
  });
  const b64: string | undefined = v?.result?.value?.data?.[0];
  if (typeof b64 !== 'string') return undefined;
  return Buffer.from(b64, 'base64');
}

function now(): number {
  return Math.floor(Date.now() / 1000);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function main(): Promise<void> {
  const endpoint = process.env.RPC_ENDPOINT;
  if (!endpoint) throw new Error('RPC_ENDPOINT');
  const altAddr = process.env.ALT_ADDRESS;
  if (!altAddr) throw new Error('ALT_ADDRESS');
  const borrowUi = Number.parseFloat(process.env.BORROW_USDC ?? '') || 500.0;
  const pollMs = Number.parseInt(process.env.POLL_MS ?? '', 10) || 800;
  const runDir = process.env.RUN_DIR ?? 'runs';
  const borrowAmount = BigInt(Math.trunc(borrowUi * 1e6));
  const cfg = pair();
  // Placeholder signer — simulate with sigVerify=false, so no keypair needed.
  const signer = new PublicKey('Anu6Awu4kxaEDrg1nkpcikx6tJ2xhfVci5TvDrZBsZEB');

  const altData = await accountData(endpoint, altAddr);
  if (altData === undefined) throw new Error('ALT');
  const alt = loadAlt(altAddr, altData);
  mkdirSync(runDir, { recursive: true });
  const out = `${runDir}/profit_watch.jsonl`;

  console.error(`profit-watch ${cfg.label} borrow ${borrowUi} USDC poll ${pollMs}ms — simulating both dirs; logs clean hits → ${out}`);
  let iters = 0n;
  let clean = 0n;
  let bestEdgeBps = Number.NEGATIVE_INFINITY;

  for (;;) {
    iters += 1n;
    const [o, r] = await Promise.all([accountData(endpoint, cfg.orcaPool), accountData(endpoint, cfg.rayPool)]);
    if (o === undefined || r === undefined) {
      await sleep(pollMs);
      continue;
    }
    // Spot edge for context (stale-free here: just-fetched pools).
    const po = orcaPrice(o);
    const pr = rayClmmPrice(r);
    let edgeBps = Number.NaN;
    if (po !== undefined && pr !== undefined && po > 0.0 && pr > 0.0) {
      edgeBps = (Math.abs(pr - po) / Math.min(po, pr)) * 1e4 - cfg.roundTripFeeBps();
    }
    if (Number.isFinite(edgeBps) && edgeBps > bestEdgeBps) {
      bestEdgeBps = edgeBps;
    }
    const pools: PoolData = { orca: o, ray: r };
    const bh = DEFAULT_BLOCKHASH;

    for (const orcaFirst of [false, true]) {
      let tx;
      try {
        tx = buildArbTx(pools, signer, alt, borrowAmount, orcaFirst, undefined, 0n, 10_000n, bh, 0n);
      } catch {
        continue;
      }
      const b64 = Buffer.from(tx.serialize()).toString('base64');
      const v = await rpc(endpoint, {
        jsonrpc: '2.0',
        id: 1,
        method: 'simulateTransaction',
        params: [b64, { encoding: 'base64', sigVerify: false, replaceRecentBlockhash: true }],
      });
      const val = v?.result?.value ?? null;
      if (val !== null && (val?.err === null || val?.err === undefined)) {
        clean += 1n;
        const dir = orcaFirst ? 'orca→ray' : 'ray→orca';
        const cu = val?.unitsConsumed;
        console.error(`🎉 CLEAN SIM [${dir}] — profitable arb exists NOW, tx would land (edge≈${edgeBps.toFixed(2)}bp, cu=${cu})`);
        const row = { t: now(), dir, edge_bps: edgeBps, cu, borrow_usdc: borrowUi };
        try {
          await appendFile(out, `${JSON.stringify(row)}\n`);
        } catch {
          // best-effort, mirrors the Rust `let _ =` swallow
        }
      }
    }
    if (iters % 50n === 0n) {
      console.error(`[profit-watch] iters=${iters} clean_sims=${clean} best_edge=${bestEdgeBps.toFixed(2)}bp (need >0 to profit)`);
    }
    await sleep(pollMs);
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
